src/main/java/com/aws/greengrass/clientdevices/auth/iot/Thing.java

```diff
@@ -8,29 +8,138 @@
 import com.aws.greengrass.clientdevices.auth.session.attribute.AttributeProvider;
 import com.aws.greengrass.clientdevices.auth.session.attribute.DeviceAttribute;
 import com.aws.greengrass.clientdevices.auth.session.attribute.WildcardSuffixAttribute;
-import lombok.Value;
+import lombok.Getter;
 
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.regex.Pattern;
 
-@Value
-public class Thing implements AttributeProvider {
+/**
+ * This is a versioned representation of an IoT Thing. It is **NOT** updated
+ * when the local Thing Registry is updated, or when changes to this Thing are
+ * made in IoT Core. Consider calling the ThingRegistry to retrieve Thing
+ * objects as they are needed rather than storing references long term.
+ */
+@Getter
+public final class Thing implements AttributeProvider, Cloneable {
     public static final String NAMESPACE = "Thing";
     private static final String thingNamePattern = "[a-zA-Z0-9\\-_:]+";
 
-    String thingName;
+    private final int version;
+    private final String thingName;
+    private final CopyOnWriteArrayList<String> attachedCertificateIds;
+    private boolean modified = false;
 
     /**
-     * Constructor.
+     * Create a new Thing.
+     *
      * @param thingName AWS IoT ThingName
      * @throws IllegalArgumentException If the given ThingName contains illegal characters
      */
-    public Thing(String thingName) {
+    public static Thing of(String thingName) {
+        return Thing.of(0, thingName);
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName) {
+        return Thing.of(version, thingName, new ArrayList<>());
+    }
+
+    /**
+     * Create a new Thing.
+     *
+     * @param version        Thing version
+     * @param thingName      AWS IoT ThingName
+     * @param certificateIds Attached certificate IDs
+     * @throws IllegalArgumentException If the given ThingName contains illegal characters
+     */
+    public static Thing of(int version, String thingName, List<String> certificateIds) {
+        if (version < 0) {
+            throw new IllegalArgumentException("Invalid version. Version must not be < 0");
+        }
         if (!Pattern.matches(thingNamePattern, thingName)) {
             throw new IllegalArgumentException("Invalid thing name. The thing name must match \"[a-zA-Z0-9\\-_:]+\".");
         }
+        return new Thing(version, thingName, new CopyOnWriteArrayList<>(certificateIds));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;certificateIds&#39; (index: 2 | type: List). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;thingName&#39;, &#39;version&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).